### PR TITLE
fix: set default value for route schema to match type

### PR DESCRIPTION
**Pull request by nflaig**
Alternative to https://github.com/fastify/fastify/pull/5394 which changes runtime behavior instead of updating type

Closes https://github.com/fastify/fastify/issues/5393

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- ~~[ ] documentation is changed or added~~
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
Can you add a test for this change?

**Comment by mcollina** [comment]
I personally prefer this to https://github.com/fastify/fastify/pull/5394.

**Comment by nflaig** [comment]
> Can you add a test for this change?

done

**Comment by nflaig** [comment]
> I personally prefer this to #5394.

yeah, I like it more as well since updating the type might annoy some people, and it's really just an edge case inside hooks

**Review by mcollina — APPROVED** [review]
lgtm

**Review by Fdawgs — APPROVED** [review]

File: lib/request.js [hunk 1]
```diff
@@ -206,7 +206,7 @@ Object.defineProperties(Request.prototype, {
           get: () => context.config
         },
         schema: {
-          get: () => context.schema
+          get: () => context.schema ?? {}
```
Comment (gurgunday) [inline lib/request.js:209]: Isn't [this](https://github.com/fastify/fastify/pull/5395/files#diff-2871801a54601791d9e1004f04b2df5d7adc881560285ff3fa6bee7bf9697857R231) enough to not get undefined?

Looks like `?? {}` here is a duplicate

```suggestion
          get: () => context.schema
```

**Review by gurgunday — APPROVED** [review]
lgtm

I think in `next` we should fix the type instead and it should be undefined

File: test/schema-feature.test.js [hunk 2]
```diff
@@ -82,6 +82,22 @@ test('The schema should be accessible by id via getSchema', t => {
   fastify.ready(err => t.error(err))
 })
 
+test('The schema should not be undefined in hooks if route does not exist', t => {
+  t.plan(4)
+  const fastify = Fastify()
+
+  fastify.addHook('onRequest', function (req, _reply, done) {
+    t.not(req.routeSchema, undefined)
+    t.not(req.routeOptions.schema, undefined)
```
Comment (Eomm) [inline test/schema-feature.test.js:91]: TBH I don't like this default.

How do the user know if a schema has been set?

Now: `if(req.routeSchema==null)`

Then: `if(req.routeSchema.id==null)` + the overhead to create a new empty object


Why don't we fix the types instead?

Comment (nflaig) [reply test/schema-feature.test.js:91]: Reopened https://github.com/fastify/fastify/pull/5394, both PRs fix the issue I reported but you guys know best what other use cases to consider.

I go with whatever is decided here :)

Comment (gurgunday) [reply test/schema-feature.test.js:91]: If we go with this in v4, I’d say right after we should revert this in v5 and merge your other (type) PR there

Comment (nflaig) [reply test/schema-feature.test.js:91]: > If we go with this in v4, I’d say right after we should revert this in v5 and merge your other (type) PR there

sounds like a plan, do you want the revert of this to be part of the type PR then? (needs rebase on next first)

Comment (mcollina) [reply test/schema-feature.test.js:91]: @gurgunday @Eomm then we go with the other PR. Let's not land and revert.

**Comment by nflaig** [comment]
Closing as https://github.com/fastify/fastify/pull/5394 has been merged to `main`
